File: java/src/org/gessinger/gepard/JSAcc.java
package org.gessinger.gepard ;

import java.util.* ;
import com.google.gson.* ;

public class JSAcc
{
  Map map = null ;
  public JSAcc()
  {
    this ( null ) ;
  }
  public JSAcc ( Map m )
  {
    if ( ! ( m instanceof Map ) )
    {
      m = new HashMap() ;
    }
    map = m ;
  }
  public String toString (  )
  {
    return Util.toString ( map ) ;    
  }
  /**
   * @brief      Get the map
   *
   * @return     the map
   */
  public Map map (  )
  {
    return map ;    
  }
  /**
   * @brief      Retrieve a value object from the JSON tree
   *
   * @param      path  '/' separated names
   *
   * @return     found object or null
   */
  public Object value ( String path )
  {
    return value ( path, null ) ;
  }
  /**
   * @brief      Retrieve a value object from the JSON tree
   *
   * @param      path  '/' separated names
   * @param      def   default object if path does not exist.
   *
   * @return     found object or given default
   */
  public Object value ( String path, Object def )
  {
    return value ( map, path, def ) ;
  }
  /**
   * @brief      Create path if it does not exist
   * 
   * @param      path  '/' separated names
   *
   * @return     { description_of_the_return_value }
   */
  public Object add ( String path )
  {
    return add ( map, path ) ;    
  }
  /**
   * @brief      Create path if it does not exist and store obj
   * 
   * @param      path  '/' separated names
   * @param      obj   object to be stared
   *
   * @return     { description_of_the_return_value }
   */
  public Object add ( String path, Object obj )
  {
    return add ( map, path, obj ) ;    
  }
  /**
   * @brief      Retrieve a value object from the JSON tree
   *
   * @param      m     Map of interest
   * @param      path  '/' separated names
   *
   * @return     found object or null
   */
  public static Object value ( Map m, String path )
  {
    return value ( m, path, null ) ;
  }
  /**
   * @brief      Retrieve a value object from the JSON tree
   *
   * @param      m     Map of interest
   * @param      path  '/' separated names
   * @param      def   default object if path does not exist.
   *
   * @return     found object or given default
   */
  public static Object value ( Map m, String path, Object def )
  {
    if ( path.indexOf ( "/" ) == -1 )
    {
      Object o = m.get ( path ) ;
      return o == null ? def : o ;
    }

    String[] plist = path.split ( "/" ) ;
    Map mm = m ;
    for ( int i = 0 ; i < plist.length ; i++ )
    {
      String p = plist[i] ;
      if ( p.length() == 0 )
      {
        continue ;
      }
      Object o = mm.get ( p ) ;
      if ( o == null )
      {
        return def ;
      }
      if ( i == plist.length - 1 )
      {
        return o ;
      }
      if ( o instanceof Map )
      {
        mm = (Map) o ;
        continue ;
      }
    }
    return def ;
  }
  /**
   * @brief      add only a map
   *
   * @param      Map root  Map
   * @param      String path  path
   *
   * @return     the new created map
   */
  public static Object add ( Map root, String path )
  {
    return add ( root, path, null ) ;
  }
  /**
   * @brief      Create path if it does not exist and store obj
   *
   * @param      root  Map of interest
   * @param      path  '/' separated names
   * @param      obj   object to be stared
   *
   * @return     { description_of_the_return_value }
   */
  public static Object add ( Map root, String path, Object obj )
  {
    if ( path.indexOf ( "/" ) == -1 )
    {
      root.put ( path, obj ) ;
      return obj ;
    }
    String[] plist = path.split ( "/" ) ;
    Map m = root ;
    for ( int i = 0 ; i < plist.length ; i++ )
    {
      String p = plist[i] ;
      if ( p.length() == 0 )
      {
        continue ;
      }
      Object o = m.get ( p ) ;
      if ( i < plist.length - 1 )
      {
        if ( ! ( o instanceof Map ) )
        {
          Map mm = new HashMap() ;
          m.put ( p, mm ) ;
          m = mm ;
        }
        if ( o instanceof Map )
        {
          m = (Map) o ;
        }
        continue ;
      }
      if ( i == plist.length - 1 )
      {
        if ( obj == null ) obj = new HashMap() ;
        m.put ( p, obj ) ;
      }
    }
    return obj ;
  }
  /**
   * @brief      remove path
   *
   * @param      path  
   *
   * @return     removed object
   */
  public Object remove ( String path )
  {
    return remove  ( map, path ) ;
  }
  static public Object remove ( Map m, String path )
  {
    if ( path.indexOf ( "/" ) == -1 )
    {
      return m.remove  ( path ) ;
    }
    String[] plist = path.split ( "/" ) ;
    Map mm = m ;
    for ( int i = 0 ; i < plist.length ; i++ )
    {
      String p = plist[i] ;
      if ( p.length() == 0 )
      {
        continue ;
      }
      Object o = mm.get ( p ) ;
      if ( o == null )
      {
        return null ;
      }
      if ( i == plist.length - 1 )
      {
        mm.remove ( p ) ;
        return o ;
      }
      if ( o instanceof Map )
      {
        mm = (Map) o ;
        continue ;
      }
    }
    return null ;
  }
  public static void main(String[] args)
  throws Exception
  {
    Gson gson = new Gson() ;
    JSAcc a = new JSAcc () ;
    a.add ( "M1/M2/N", 11 ) ;
// System.out.println ( "1 ----------------------" ) ;
  System.out.println ( gson.toJson ( a.value ( "M1/M2/N" ) ) ) ;
  System.out.println ( gson.toJson ( a.value ( "M1/M2" ) ) ) ;
    a.add ( "A/B/C", new String[] { "ABCD", "ABCE" } ) ;
    a.add ( "A/B/D", "ABCD" ) ;
    a.add ( "A/AX", "AX" ) ;
    a.add ( "X", "X" ) ;
// System.out.println ( "2 ----------------------" ) ;
  System.out.println ( gson.toJson ( a.map ) ) ;
// System.out.println ( "3 ----------------------" ) ;
  System.out.println ( gson.toJson ( a.value ( "A/B" ) ) ) ;
    a.remove ( "X" ) ;
// System.out.println ( "4 -- a.remove ( 'X' )" ) ;
  System.out.println ( gson.toJson ( a.map ) ) ;
    a.remove ( "A/B/D" ) ;
// System.out.println ( "5 -- a.remove ( 'A/B/D' )" ) ;
  System.out.println ( gson.toJson ( a.map ) ) ;
    a.remove ( "A" ) ;
// System.out.println ( "6 -- a.remove ( 'A' )" ) ;
  System.out.println ( gson.toJson ( a.map ) ) ;
    a.remove ( "M1/M2/N" ) ;
// System.out.println ( "7 -- a.remove ( 'M1/M2/N' )" ) ;
  System.out.println ( gson.toJson ( a.map ) ) ;
    a.add ( "only/a/map" ) ;
  System.out.println ( gson.toJson ( a.map ) ) ;
  }
}

File: src/JSAcc.js
#!/usr/bin/env node

var util = require ( "util" ) ;

var JSAcc = function ( map )
{
  this.className = "JSAcc" ;
  if ( typeof map === 'object' )
  {
    this.map = map ;
  }
  else
  {
    this.map = {} ;
  }
};
JSAcc.prototype.toString = function()
{
  return util.inspect ( this.map, { showHidden: false, depth: null } ) ;    
};
JSAcc.prototype.map = function()
{
  return this.map ;    
};
JSAcc.prototype.value = function ( path, def )
{
  if ( path.indexOf ( "/" ) == -1 )
  {
    var o = this.map[path] ;
    return typeof o === 'undefined' ? def : o ;
  }

  var plist = path.split ( "/" ) ;
  var mm = this.map ;
  for ( var i = 0 ; i < plist.length ; i++ )
  {
    var p = plist[i] ;
    if ( ! p )
    {
      continue ;
    }
    var o = mm[p] ;
    if ( typeof o === 'undefined' )
    {
      return def ;
    }
    if ( i === plist.length - 1 )
    {
      return o ;
    }
    if ( typeof o === 'object' )
    {
      mm = o ;
      continue ;
    }
  }
  return def ;
};
JSAcc.prototype.add = function ( path, obj )
{
  if ( path.indexOf ( "/" ) == -1 )
  {
    this.map[path] = obj ;
    return obj ;
  }
  var plist = path.split ( "/" ) ;
  var m = this.map ;
  for ( var i = 0 ; i < plist.length ; i++ )
  {
    var p = plist[i] ;
    if ( ! p )
    {
      continue ;
    }
    var o = m[p] ;
    if ( i < plist.length - 1 )
    {
      if ( typeof o !== 'object' )
      {
        var mm = {} ;
        m[p] = mm ;
        m = mm ;
      }
      if ( typeof o ==='object' )
      {
        m = o ;
      }
      continue ;
    }
    if ( i === plist.length - 1 )
    {
      m[p] = obj ;
    }
  }
  return obj ;
};
JSAcc.prototype.remove = function ( path )
{
  if ( path.indexOf ( "/" ) == -1 )
  {
    var o = this.map[path] ;
    delete this.map[path] ;
    return o ;
  }
  var plist = path.split ( "/" ) ;
  var mm = this.map ;
  for ( var i = 0 ; i < plist.length ; i++ )
  {
    var p = plist[i] ;
    if ( ! p )
    {
      continue ;
    }
    var o = mm[p] ;
    if ( typeof o === 'undefined' )
    {
      return ;
    }
    if ( i == plist.length - 1 )
    {
      delete mm[p] ;
      return o ;
    }
    if ( typeof o === 'object' )
    {
      mm = o ;
      continue ;
    }
  }
  return ;
};
module.exports = JSAcc ;
if ( require.main === module )
{
  var T = require ( "./Tango" ) ;
  var a = new JSAcc () ;

  a.add ( "M1/M2/N", 11 ) ;
// console.log ( '1 ----------------------' ) ;
  console.log ( JSON.stringify ( a.value ( "M1/M2/N" ) ) ) ;
  console.log ( JSON.stringify ( a.value ( "M1/M2" ) ) ) ;
  a.add ( "A/B/C", [ "ABCD", "ABCE" ] ) ;
  a.add ( "A/B/D", "ABCD" ) ;
  a.add ( "A/AX", "AX" ) ;
  a.add ( "X", "X" ) ;
// console.log ( '2 ----------------------' ) ;
  console.log ( JSON.stringify ( a.map ) ) ;
// console.log ( '3 ----------------------' ) ;
  console.log ( JSON.stringify ( a.value ( "A/B" ) ) ) ;
  a.remove ( "X" ) ;
// console.log ( '4 -- a.remove ( "X" )' ) ;
  console.log ( JSON.stringify ( a.map ) ) ;
  a.remove ( "A/B/D" ) ;
// console.log ( '5 -- a.remove ( "A/B/D" )' ) ;
  console.log ( JSON.stringify ( a.map ) ) ;
  a.remove ( "A" ) ;
// console.log ( '6 -- a.remove ( "A" )' ) ;
  console.log ( JSON.stringify ( a.map ) ) ;
  a.remove ( "M1/M2/N" ) ;
// console.log ( '7 -- a.remove ( "M1/M2/N" )' ) ;
  console.log ( JSON.stringify ( a.map ) ) ;
}
